CoverBrowser plugin: alt views for File Browser and History

poire-z opened the pull request:
This plugin offers some alternative views of file lists as used by FileManager and History, using "as you browse" extracted metadata and cover images from files.
It exports one new sub-menu (Display mode) to File Browser menu and upon selected mode, replace a few display methods of the FileChooser and FileManagerHistory classes. While being a bit hacky, it has the effect that all the other code of these is still used as is and all features available (sorting, hidden files, select and hold actions...), so very few code duplicated.
As of now, it does not need any modification to core code, and is fully self-contained in its plugin directory.
 
<kbd>![1-menu-textcover](https://cloud.githubusercontent.com/assets/24273478/26780997/981c3344-49ec-11e7-8b52-c80fbb95319d.png) </kbd>
(If you choose "Classic" mode, this plugin does nothing.)

<kbd>![2-browser-coverimages](https://cloud.githubusercontent.com/assets/24273478/26781000/9b57b2d6-49ec-11e7-8573-25b2314c273e.png)</kbd>

<kbd>![3-browser-listimages](https://cloud.githubusercontent.com/assets/24273478/26781004/9fdb0e8e-49ec-11e7-8de4-92cd08bf80ea.png)</kbd>

<kbd>![4-history-coverimages](https://cloud.githubusercontent.com/assets/24273478/26781007/a267faae-49ec-11e7-930a-148529015fe0.png)</kbd>

What's your feeling about this, a plugin dynamically replacing core code (and putting it back when you choose Classic again) ?
(Could have been named koverbrowser, K-Over-browser :)


Long story ... long :

Got interested by this comment by @Frenzie https://github.com/koreader/koreader/issues/2906#issuecomment-302744109 and wondered how we could do that with our single threaded KOReader.
Thought about having another code starter (indexer.lua, like reader.lua), but it got complicated to re-initialize the whole reader stuff to the point of having the engines working.
Turned out the easiest and cheapest way to have indexing in the background was just to (unix-)fork the main reader process, have it do the opening of documents, get the metadata and cover image, save them, and die. And it works beautifuly (on kobo, should be fine on any linux based device).  Haven't seen any side effect of doing that. A bit of subprocess management and collecting, and it got really safe and sound.
(It's disabled for Android, as I don't know if fork works, and what are the alternatives.)

I didn't want any .sdr directory to be created for each file to hold a thumbnail of the cover.
So I choose SQLite to store all these informations (I learnt the lua-ljsqlite3 way of things by looking at @robert00s 's Stats plugin v2, thanks!).
It's fine as it's really fast, and deals with integrity and keeping concurrent accesses by main process (for reading) and indexer sub-process (for writing) in queue.
Cached book informations are just in one file (`settings/bookinfo_cache.sqlite3`), and it can be deleted as it's just a cache, so it's quite clean. It may get big, 5MB for a directory with 40 books if you choose Mosaic images, smaller for other modes, but it's your choice. (Not done yet: cleanup of removed files.)
I also choose to store in there, not a jpeg or png version of the cover thumbnail, but a zlib-compressed dump of a blitbuffer sized exactly as the thumbnail that will be displayed. That makes file browsing really smooth. As soon as there is re-scaling involved, you notice the loss in smoothness. Should be even worse if a mupdf rendering of a png to blitbuffer was needed.

So, as you browse, and you're shown a 3x3 grid of not yet indexed books, a background process is launched to open these 9 books and extract stuff and add them to the SQLite db (an entry is created for each full path, even if the book as no metadata or cover).
In the meantime, a schedule action watches every second if the entry for each book is in the db, and if it is, re-display the book with the new info and cover (or alternate representation if no metadata or no cover was found in the book).
Next time you revisit the page with these 9 books, their entries are there in the db and used, so no more background indexing needed.  So, as you browse, you're indexing only the parts of the filesystem you're looking at.
Safe guards here and there to stop subprocesses when you change page, quit filebrowser, or fall asleep.

Building the differents UI elements was fun, with the satisfaction of visual and animated feedback :)
I read discussions and used suggestions from #2614 and #2906.  Although it's not a real "home", it's just alternative views for FileManager and History, which in the end may fit most of the wishes in these discussions, and don't cause as much objections (including mine:).
(I have no interest in a global metadata and tags db, and browse/sort/seach by metadata.  I also don't want to contribute in any way to the decline of the art of organizing by folder :)
I was just curious to see how books would look as thumbnails, and it turned out the title & authors metadata where quite good in the many books of different origins I have, and good enough for display.

Added a few more buttons to the onHold popup so people can ignore some of these if they are bad or offending :

![5-onhold](https://cloud.githubusercontent.com/assets/24273478/26781394/889054a8-49ee-11e7-97f9-71928306c28e.png)

That was the fun part.

The moderately fun part was to learn all the widgets toolkit, containers, groups and how they work :)

The less fun part that took me too much time was chasing memory leaks (ImageWidget blitbuffers, and TextWidget/TextBoxWidget too have blitbuffers !), and make sure to call free() where logically needed.
And the half-pixel positionning ! I've seen some `paintTo(bb, 18, 124.5)` that caused some misalignment of images in their bordered framecontainer, eating the top or bottom border (the same code was fine on the emulator, fine in FileBrowser on my Kobo, but bad in History ; all that because on my Kobo, the "History" header TextWidget had `height=18.5` , and so all the below positionning had a `y=NNN.5`. Hours of cursing containers widgets before finding that :) Comments in the code about how to fix this in TextWidget, if that's the right thing to do.)

Some comments in the code about stuff to add/fix/move to core or base. Welcoming feedback.

Frenzie commented:
> It exports one new sub-menu (Display mode) to File Browser menu and upon selected mode, replace a few display methods of the FileChooser and FileManagerHistory classes. While being a bit hacky, it has the effect that all the other code of these is still used as is and all features available (sorting, hidden files, select and hold actions...), so very few code duplicated.

That is rather brilliant as a way to test-drive and finetune changes even if intended for later integration with the core.

> (I have no interest in a global metadata and tags db, and browse/sort/seach by metadata. I also don't want to contribute in any way to the decline of the art of organizing by folder :)

\*grins\*

Frenzie commented on plugins/coverbrowser.koplugin/covermenu.lua:
```diff
@@ -0,0 +1,383 @@
+            -- on screen, so we won't see it)
+            UIManager:close(self.file_dialog)
+
+            -- Add some new buttons to orignal buttons set
```
Stupid nitpick: original

Frenzie commented on plugins/coverbrowser.koplugin/covermenu.lua:
```diff
@@ -0,0 +1,383 @@
+    -- Add some new buttons to orignal buttons set
+    table.insert(orig_buttons, {
+        { -- Allow user to ignore some offending cover image
+            text = bookinfo.ignore_cover and _("Unignore file cover") or _("Ignore this cover"),
```
I think just `Ignore cover` would do the trick.

poire-z replied:
Do you want that too for the Unigore entries ?

Frenzie replied:
Yeah. :-)

Frenzie commented on plugins/coverbrowser.koplugin/covermenu.lua:
```diff
@@ -0,0 +1,383 @@
+            end,
+        },
+        { -- Allow user to ignore some bad metadata (filename will be used instead)
+            text = bookinfo.ignore_meta and _("Unignore file metadata") or _("Ignore these metadata"),
```
Same for `Ignore metadata`. In fact "Ignore these metadata` has an odd ring to it.

Frenzie commented on plugins/coverbrowser.koplugin/listmenu.lua:
```diff
@@ -0,0 +1,686 @@
+-- ItemShortCutIcon (for keyboard navigation) is private to menu.lua and can't be accessed,
+-- so we need to redefine it
+local ItemShortCutIcon = WidgetContainer:new{
+    dimen = Geom:new{ w = 22, h = 22 },
```
I think you're missing some scaling here?

poire-z replied:
That's original code cut and pasted from frontend/ui/widget/menu.lua.
(To show keyboard keys, as seen on the emulator - I don't know if there's many devices that uses keyboard, and if these are high resolution and would need scaling - Nearly got rid of this ItemShortCutIcon, but thought I stay as close to the original menu code)

Frenzie replied:
Ah, there must be some hidden scaling built into the widget then.

![screenshot_2017-06-05_15-22-13](https://cloud.githubusercontent.com/assets/202757/26785457/c6860398-4a02-11e7-8c2b-416fde41a3d4.png)

Frenzie commented on plugins/coverbrowser.koplugin/listmenu.lua:
```diff
@@ -0,0 +1,686 @@
+local ItemShortCutIcon = WidgetContainer:new{
+    dimen = Geom:new{ w = 22, h = 22 },
+    key = nil,
+    bordersize = 2,
```
Scaling.

Frenzie commented on plugins/coverbrowser.koplugin/listmenu.lua:
```diff
@@ -0,0 +1,686 @@
+    end
+    local sc_face
+    if self.key:len() > 1 then
+        sc_face = Font:getFace("ffont", 14)
```
I recently unified the various custom font sizes around the app because it was getting out of hand. Your custom definition here is awful close to the standard `smallffont` I defined based on what was spread around the code. https://github.com/koreader/koreader/blob/e1aa57f27ef41c314ff2f0ee3cc05942a5645bf1/frontend/ui/font.lua#L61

poire-z replied:
(That too is original menu.lua code, not mine.)
For the other coming font definitions you're going to see, I initially started by using picks from these set of standard fonts. Had to adjust some for the layout, and went back to using the standard "cfont" with the size that made each look fine, and fit in the hardcoded scaleBySize(64).

Frenzie replied:
I must've ignored those as a kind of special case or to deal with later when I had more time (i.e., not quite yet) and forgot about them.

Frenzie commented on plugins/coverbrowser.koplugin/mosaicmenu.lua:
```diff
@@ -0,0 +1,709 @@
+            if not loop2  then
+                loop2 = true
+                sizedec = self.initial_sizedec -- restart from initial big size
+                -- Replace underscores and hyphens with spaces, to allow text wrap there.
```
Wait, we don't wrap on hyphens? Kind of sounds like that should probably be a thing. :-P

Frenzie commented on plugins/coverbrowser.koplugin/bookinfomanager.lua:
```diff
@@ -0,0 +1,492 @@
+
+        -- File location and filename
+        directory           TEXT NOT NULL, -- split by dir/name so we can get all files in a directory
+        filename            TEXT NOT NULL, -- and can implement pruning of no more existing files
```
This does kind of illustrate why I brainstormed about using the .sdr directory: it moves along with the book without any extra effort (and you can easily effect that manually on USB).

poire-z replied:
Yes, but you have to think about not forgetting to move the .sdr too on USB.
With this sqlite db, if you move a file or rename a folder, you just have obsolete cache data in sqlite. Nothing really needed to move along. And new cached data will be regenerated on demand.
But that will make for another item in DB management sub menu: Prune obsolete info (better idea for that button title ?)

Frenzie replied:
That does sound like a better description, but I'm not really sure if that's where it belongs. It seems more like another action for the book info cache manager.

poire-z replied:
Sorry, not button: menu entry below Book info cache management / Purge cache (not yet implementing it, just added it as TODO comment)

Frenzie commented:
There's nothing that jumps out at me on this one. Just a couple of nitpicky things like a stray newline here and there and the things I already commented.

tl;dr Lookin' good!

Frenzie commented on plugins/coverbrowser.koplugin/bookinfomanager.lua:
```diff
@@ -0,0 +1,492 @@
+    self.subprocesses_pids = {}
+    self.subprocesses_last_added_ts = nil
+    self.subprocesses_killall_timeout_seconds = 300 -- cleanup timeout for stuck subprocesses
+        -- 300 seconds should be enough to open and get info from 9-10 books
```
Some stray indentation.

Frenzie commented on plugins/coverbrowser.koplugin/main.lua:
```diff
@@ -0,0 +1,353 @@
+                    },
+                },
+                {
+                    text = _("Book infos cache maintenance"),
```
`Book info cache maintenance`

Info/information is a mass noun like water, sand or electricity. You can't count it or pluralize it.

Also it might be better to call it *management* instead of *maintenance*?

Frenzie commented on plugins/coverbrowser.koplugin/main.lua:
```diff
@@ -0,0 +1,353 @@
+        display_mode = nil
+    end
+    self.filemanager_display_mode = display_mode
+    G_reader_settings:saveSetting("filemanager_display_mode", self.filemanager_display_mode)
```
Incidentally, while I don't exactly perceive it as a problem in this case because I feel that this is perhaps more of a kind of experimental core expansion than a plugin per se, you are effectively polluting the global settings with plugin-specific settings here.

So from my end this is more of a "be aware that this is what you're doing" than a "please change it".

Frenzie commented on plugins/coverbrowser.koplugin/main.lua:
```diff
@@ -0,0 +1,353 @@
+
+    -- We do now the single thing done in FileManagerHistory:updateItemTable():
+    hist_menu:switchItemTable(self.hist_menu_title, require("readhistory").hist)
+
```
Stray newline.

Frenzie commented on plugins/coverbrowser.koplugin/xutil.lua:
```diff
@@ -0,0 +1,103 @@
+local zlib = ffi.load(ffi.os == "Windows" and "zlib1" or "z")
+
+function xutil.zlib_compress(data)
+  local n = zlib.compressBound(#data)
```
bunch of two-space indentation, same in `xutil.zlib_uncompress`

KenMaltby commented:
Can I safely assume that if "Classic (filename only)" is selected no SQLite data base will be generated and no processing to support it activated?

poire-z commented:
You can. You could even test the new displays, browse whatever you want, get back to classic, and choose DB management / Purge cache db, and the db would be deleted - and you'd be as if you never selected any of the new displays.

poire-z commented:
Thanks. Yes, i'm now aware of the 2 global settings, but as you hint, as they are about core stuff (filemanager, history) behaviour, they may find their home there with all the other core settings.

Frenzie commented on plugins/coverbrowser.koplugin/covermenu.lua:
```diff
@@ -0,0 +1,380 @@
+
+local BookInfoManager = require("bookinfomanager")
+
+
```
newline

Frenzie commented on plugins/coverbrowser.koplugin/covermenu.lua:
```diff
@@ -0,0 +1,380 @@
+-- not found item to self.items_to_update for us to update() them
+-- regularly.
+
+
```
another

Frenzie commented on plugins/coverbrowser.koplugin/covermenu.lua:
```diff
@@ -0,0 +1,380 @@
+-- in the real Menu class or instance
+local CoverMenu = {}
+
+
```
more :-P

Frenzie commented on plugins/coverbrowser.koplugin/covermenu.lua:
```diff
@@ -0,0 +1,380 @@
+    collectgarbage()
+    collectgarbage()
+
+
```
newline here and a few lines below as well

Frenzie commented on plugins/coverbrowser.koplugin/covermenu.lua:
```diff
@@ -0,0 +1,380 @@
+    if self.path_items then
+        self.path_items[self.path] = (self.page - 1) * self.perpage + (select_number or 1)
+    end
+
```
newline

Frenzie commented:
Have you thought about unit tests? There are the test files in `/test` so I think it would be a good idea to add some for `BookInfoManager:getBookInfo()` if nothing else. Especially for potential failure scenarios.

poire-z commented:
Can't 2 newlines sometimes make sense ? :) I usually use them to separate functions related to different stuff (like bookinfo db vs subprocess management in bookinfo.lua), or to emphasis a change of kind of work in the same function.
I'm not used at making unit tests, so may be later, once it's accepted or stable.

Btw, lua check reported a warning (unused variable 'directory') for this:
`local directory, filename = splitFilePathName(filepath)`
(that I used at 4 places, but it reported it only at the place where I wasn't needing or using directory.
Had to change it after a few minutes of pondering to
`local filename = select(2, util.splitFilePathName(self.filepath))`
which is less clear.
Was there an other solution ?

Frenzie commented:
It could make sense, but if so it should probably be added to https://github.com/koreader/koreader-base/wiki/Coding-style But that's a separate discussion.

Regarding luacheck, try either `_, filename` or something like this:

```lua
local directory, filename = splitFilePathName(filepath) -- luacheck: ignore
```

poire-z commented:
OK, i'll put back the clearer original line and use `-- luacheck: ignore`.
I couldn't use _ as a few lines later in the same scope, I needed to use `_("something")`...

apletnev commented:
@poire-z looks awesome. thanks. Only one question, did you check how it looks like in landscape mode? For example, I use my reader only in landscape mode, and unfortunately, some UI components are hardcoded only for portrait mode. Thank you again, nice work.

poire-z commented:
Yes I did. I made it a 4x2 grid if Screen.getWidth() > Screen.getHeight() instead of the 3x3, and it looks as nice. (Although I often had a hard time getting FileBrowser to display in landscape mode, even when document was in landscape mode (original code, not this plugin code I guess) - sometimes it did, sometimes not - but when it did, it allowed me to check it was nice).

Frenzie commented:
I'll wait a little longer for someone else to take a peek before merging.

poire-z commented:
No hurry from my part.
I'd like @houqp to have a look, and his opinion about this plugin plugging (uh!) that way into core, and that it doesn't go against some other plan about metadata/home.
Also, there's some bits and fixes in this plugin that may probably find their place in core or base first, so I was expecting to have a small rewrite to do :)
I made it self-contained so people can easily test it (have you tried it ?) I'd like to have feedback about if/how it works on kindle or pocketbook. Should I provice a zip to unzip in plugin/ dir ?

Frenzie commented:
I only have an H2O. Seems to work okay.

On Thu, Jun 8, 2017 at 11:28 PM, poire-z <notifications@github.com> wrote:

> No hurry from my part.
> I'd like @houqp <https://github.com/houqp> to have a look, and his
> opinion about this plugin plugging (uh!) that way into core, and that it
> doesn't go against some other plan about metadata/home.
> Also, there's some bits and fixes in this plugin that may probably find
> their place in core or base first, so I was expecting to have a small
> rewrite to do :)
> I made it self-contained so people can easily test it (have you tried it
> ?) I'd like to have feedback about if/how it works on kindle or pocketbook.
> Should I provice a zip to unzip in plugin/ dir ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/koreader/koreader/pull/2940#issuecomment-307203670>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAMYBQjkCa9qjyoOrHDqu7NCyh49J4Tsks5sCEtsgaJpZM4Nv8No>
> .
>

robert00s commented:
Great plugin. Thanks.
On Kindle PW3 works very well.

slow-mo commented:
I'm testing it on a Kindle PW 2 and works fine. It's very nice!
From a graphical point of view I'd suggest using a slight spacing between cover images in list view instead of dividing lines.

bv217 commented:
Thanks, great plugin. But there are some remarks.
It would be nice to display the series number of the book, if it exists.
Read and unread books do not differ, that is, part of the functionality is lost.
This book plugin could not see: https://www.load.to/9qyeXI6KtY/worm.fb2.zip
Kindle PW

poire-z commented:
> Read and unread books do not differ

Read books (opened books that have a .sdr dir) are shown with a corner mark at their bottom right, in both modes, as shown in the screenshots. Thought that was a obvious hint, but seems not :)

> It would be nice to display the series number of the book

I could add an option to append them to title or authors. Is there a consensus about which of these two to append Series to (also note that from code, the single `Series` metadata includes Series name + number, not the number as a single metadata, so it would have to be both. Or use Series in place of Authors ?

I have no problem with your book on the emulator, this books has just two metadata: title and authors, and no cover. And the plugin displays title and authors fine.

kapaka commented:
Thanks for developing this plugin.
I would like to be able to sort books by author and then series  and number in the series inside every folder. Some series are complicated by being written by different authors and/or collaborations. Or spinoff series. One good example is the Kurtherian Gambit series, it has all the above. maybe you could use this structure to check how to display metadata. Star Wars series or Shadowrun series books have the same problem. Thank you

julianopaz commented:
Anyone can help me to intall this plugin on my stable koreader kindle ppw3?

poire-z commented:
You can go to https://github.com/poire-z/koreader/tree/coverbrowser , then click on the green button "Clone or download" and Download Zip. You'll get a full koreader source zip, but **don't unzip it** to ur Kindle (it's sources, not compiled)! Just get the plugin directory `koreader-coverbrowser\plugins\coverbrowser.koplugin` , and put it (and its 6 files) into your Kindle `koreader/plugins/` directory.

If your stable is the current stable koreader-nightly-20170612, it should work (as I made it on 20170605, with small fixes since). But download it quickly, I may push new commits soon that will make it incompatible with this last stable.

julianopaz commented:
Very thank you!

poire-z commented:
This commit needs a nightly >= 1082-gdffa2f2

The 2 display_mode settings are no more stored in the global `settings.reader.lua` , they are now stored in this plugin own db (`config` table).
(People who tried this plugin so far and like things tidy may remove the `filemanager_display_mode` and `history_display_mode` keys from their `settings.reader.lua` .)

Added 2 columns for keywords and description to bookinfo database.

Allow direct access to description and cover image from the onHold popup (buttons enabled only if there is a description or a cover):

<kbd>![onhold2](https://user-images.githubusercontent.com/24273478/28246744-4c73fe54-6a21-11e7-91a4-cea31f6b31a1.png)</kbd>


Added a `Other settings` submenu (settings also saved in the db config table), with:

<kbd>![othersettings1](https://user-images.githubusercontent.com/24273478/28246750-767e024e-6a21-11e7-83be-23d6617ff442.png)</kbd>

Including a small visual hint if book has a description (some kind of page mark at top right of book cover, I made it quite small so it does not disctract when viewing the books list, but should be noticable when one focus his eyes on a book - i tried a few other variations: shadow, long mark on top, but small mark on right looked best to me).

Including adding Series metadata to title or authors (or both, for show, in this screenshot):

<kbd>![othersettings2](https://user-images.githubusercontent.com/24273478/28246769-009425e4-6a22-11e7-9177-c56e77104862.png)</kbd>

For people who requested stuff about Series metadata, I hope this addition of it to title and authors may help (I don't want to add a 3rd line, as it would make the widget building and drawing more complicated).

Frenzie commented on plugins/coverbrowser.koplugin/main.lua:
```diff
@@ -159,12 +160,65 @@ function CoverBrowser:addToMainMenu(menu_items)
+                            end,
+                        },
+                        {
+                            text = _("Show hint for opened books in History"),
```
I don't think there's a reason to capitalize history here?

poire-z replied:
Thought that as the history widget is titled "History", that made it clearer it was this widget, and not just any history :)

Frenzie replied:
I think just lowercase is fine and preferable, but otherwise `"History"`.

Frenzie commented on plugins/coverbrowser.koplugin/main.lua:
```diff
@@ -159,12 +160,65 @@ function CoverBrowser:addToMainMenu(menu_items)
+                            end,
+                        },
+                        {
+                            text = _("Append Series metadata to Title"),
```
Also odd capitalization

poire-z replied:
Same here, to highlight Series and Titles, just like they appear in Bookinfo kvpage.
Or I could put them in single or double quotes ?
Or just as you said, in lowercase.
Just tell me after you read my reasoning, and I'll do as you prefer.

Kwuik commented:
First of all thanks for the work on this plugin. It looks cool and adds the often missed browse library with cover art show feature... I love it. 
Comments:
1st: it gets annoying that I have to go to menu - > settings -> page 2 -> new: display mode ->  whatever mode I want... Isnt there a faster way to switch layouts or to get in this menu? 
2nd: It would be nice if in the List with image one could actually choose the image size...

The only other thing missing is native USB-Mount support without leaving koreader (Kobo Glo)... But thats another topic...

Frenzie commented:
@Kwuik 

> The only other thing missing is native USB-Mount support without leaving koreader (Kobo Glo)... But thats another topic...

See #2597.

poire-z commented:
> 1st: it gets annoying that I have to go to menu - > settings -> page 2 ...

The only thing changed to koreader's core is actually moving this menu item at top of first menu, see https://github.com/koreader/koreader/pull/2940/files#diff-5c2c491f25a859558eefa5a7ae260005R9 . If you got just the plugin directory, you missed that, so you may want to add this single line manually.
Also, one may play with layouts at first, and it may get annoying changing modes, but after a short while, one will probably have decided which one to use. Not sure it needs more way of cycling thru display modes.

> 2nd: It would be nice if in the List with image one could actually choose the image size...

This is tricky, as cover images are saved in db as exactly the size of the thumbnail. Changing the grid or line numbers would require invalidating cached images, or rescaling badly sized images that would make browsing less smooth.
Anyway, for mosaic modes, you can see for yourself how it looks with a different sized grid (purge the db for higher quality thumbnails and smoother browing between changes) by uncommenting and editing:
https://github.com/poire-z/koreader/blob/f7df72eb18b5a4af160fb15552ee79a1e0098df4/plugins/coverbrowser.koplugin/main.lua#L349-L353
For list modes, you may try changing the hardcoded 64 (which decides the height of a list item, so the height of its image) at https://github.com/poire-z/koreader/blob/f7df72eb18b5a4af160fb15552ee79a1e0098df4/plugins/coverbrowser.koplugin/listmenu.lua#L653 . But it may need font sizes to be changed too to fill the increased height...

About "Go to letter" : that wouldn't fit this plugin simple idea of just overriding display methods of the FileBrowser class. Proper implementation should be in the core FileBrowser or Menu code (that already deal with changing pages, guessing current page number and total nb of pages, dealing with taps on << < p/P > >>), so that Classic mode and this plugin mode can all benefit from it.

poire-z commented:
Well, I haven't had any more idea of stuff to add to that for weeks, so I consider it stable from my side :)
May be it could be merged ?
(btw, no nightly build for the last 3 days, and gitlab looks synced)

Frenzie commented:
I'm guessing it might have to do with GL itself as the koreader.rocks server seems to be fine. They've been having some issues these past few days.

https://twitter.com/gitlabstatus/status/897375236269760513
https://gitlab.com/gitlab-com/infrastructure/issues/2529
https://gitlab.com/gitlab-com/infrastructure/issues/2539

Btw I just manually restarted the builds so you could always download the artifacts yourself (once they're done) but I'm not sure if it'd manage to upload them to the server that way.

poire-z commented:
Thanks ! Builds went fine, and v2015.11-1131-g3b5cd4c (including this plugin) is available for download at the usual nightly url) !

Frenzie commented:
As an aside, I also fixed the mimetype (see #2955) so it won't load the archives as plain text anymore.

gerroon commented:
Can we have it enabled for Android?

poire-z commented:
It would need some android specific work (that I can't do myself, as I don't know much about Android), mainly to be able to fork() current process to just execute some lua function, as done there:
https://github.com/koreader/koreader/blob/master/plugins/coverbrowser.koplugin/xutil.lua#L11
(+ the 2 next functions to know if it's done, and to terminate it).

There's already android code to execute another command:
https://github.com/koreader/android-luajit-launcher/blob/master/assets/android.lua#L1397
so, it may be possible to implement that kind of fork() via JNI, althought I don't know if android java/JNI and process/security model really allow that.

gerroon commented:
@poire-z 

Thanks for the follow up, I see that it will take time to get this feature on Android.

skygragon commented:
Do we have a ticket to address the effort for Android? This is a great feature indeed.

Frenzie commented:
Did you miss #3264? :-)

I'll have to do perform some checks on it, but it seems to be working fine.

anarcat commented:
does this PR fix the UI/UX design issues outlined in #2614?

cramoisi commented:
@poire-z  when I finish a book, i like it to stay that way (completed) but I often go back inside them for quick reference or whatever (like re-reading a marked page). But then, the reading counter reset to the percent_finished indicated in the metadata.lua.

Would it be complicated to add completed from this same metadata when the book is finished ? (one can always reset it by purging sdr even if it's a drastic way to do it)

    ["summary"] = {
        ["status"] = "complete",
        ["modified"] = "2018-02-25"
    },

poire-z commented:
Easy to do (and done), if that's ok for all people: if status is `complete` or `abandoned` (otherwise, it's `reading` or empty string), display that instead of the read %  (I don't use that book status toggler, so I would keep seeing percentages and that would not bother me :):

<kbd>![status](https://user-images.githubusercontent.com/24273478/36902558-35b17098-1e2b-11e8-967d-6dff2e3957d9.png)</kbd>


I use the same intl strings as in widget/bookstatus.lua:

<kbd>![complete](https://user-images.githubusercontent.com/24273478/36902639-7ead9bc8-1e2b-11e8-8fa0-30642e226e23.png)</kbd>

Shouldn't it be Complete**d** , to fit with Abandoned ?

> one can always reset it by purging sdr even if it's a drastic way to do it)

Or just go back to Book status, and toggle it back to `Reading`.

Frenzie commented:
I think `Finished` might be better in that case. I think having completed a book is probably more about writing than about reading.

Btw, `Dropped` might also work instead of `Abandoned`.

poire-z commented:
I agree about `Finished`, but less about `Dropped`, just because it may have more technical meanings than `Abandoned` :)
No problem with keeping the sidecar values
`local args = { "complete", "reading", "abandoned" }`
while changing the displayed strings away from them ?
`        toggle = { _("Finished"), _("Reading"), _("Abandoned") },`
Ok, making a PR.

Frenzie commented:
I think that's an an argument in favor, not against. If Latinate abandon were the more natural word then little [Bobby Tables](https://www.xkcd.com/327/) would probably be named `ABANDON TABLE`. ;-)

KenMaltby commented:
I guess "Unfinished" might be a more neutral term.  Dropped and Abandoned carry a slight negative to the reason the book wasn't finished.  (Might be a valid comment though.)

Frenzie commented:
> May be :) But you can't undropped a DROPed TABLE, while you can resume an Abandonned book (that I have a hard time writting with a single n).

You can also pick a dropped book back up. Ultimately a specific technical sense holds no sway one way or the other.

@KenMaltby How about `Gave up` xD

More seriously, `On hold`? `Shelved`?

cramoisi commented:
Great ! So Great ! Thanks !!

@KenMaltby for "unfinished" , it's the case of all books on "reading" status, so it's perhaps pretty confusing

@Frenzie @poire-z in fact, there are two kind of abandoned books, those for which we know we will come back because it's great literature ; and those which are trash (those, in fact, we delete them). So, I tend for "on hold" but in fact, these kind of books need most of the time to be starting again from page 1 (so "on hold" doesn't fit right).

Yes, I think about you War and Peace :)

I love gave up, in fact

KenMaltby commented:
I wonder if it would be seen in the listing, anyway.  Wouldn't the percentage read be displayed?

cramoisi commented:
@KenMaltby it's exactly what @poire-z did (cf. screenshot)

Frenzie commented:
@cramoisi I started reading *Pietr-le-Letton* a couple of years ago. I rather doubt it's great literature, but my French was insufficiently developed. I had to start with *Spirou*. I read all of them. The first 20 or 30 or so I'd already read in Dutch years ago, so the level was just about right. Now I read François Sagan (I don't like her protagonists but I love her writing), Pascal Quignard (a bit weird but very enjoyable), Marguerite Yourcenar (very impressive), Patrick Modiano (loved Dora Bruder, didn't care for Voyage de noces), and my favorite living author is probably Laurent Binet.

tl;dr I think I'm probably ready to return, but Pietr-le-Letton was on hold due to me not exercising sufficient command of the language. :-P

KenMaltby commented:
And that works for "Abandoned" or "Dropped", as that implies that you would not be returning to the book and would not need to know how far you got into it.

Frenzie commented:
Truly abandoned doesn't really make much sense to me. That's just deleted. /shrugs

I like "on hold". :-)

cramoisi commented:
Pietr-le-letton is great literature on its genre. And I assume you'll start it again from page 1

KenMaltby commented:
Perhaps "on hold @ " "percent read" ?

cramoisi commented:
and for the french translation, Lu ? (it's quite short, I don't know if I like it, even if I named a directory with it (1.bestof 2.lus)

cramoisi commented:
Now, we need a easier way to display the "status page" (the only ways now is to be in the book,  hamburger > book status). What about in the long press menu in the filemanager ?

KenMaltby commented:
You can "Use book status as screensaver" which would give you the current books status page.  For those books in "History" you might add such a "long press" option (to the already long list of options).  `For the books/files in the file browser, you would have to generate the page to display, I don't know if that would be costly or not.

poire-z commented:
Having Book status available in the onHold menu would be a bit harder to do, as this widget currently expects a ReaderUI with that document opened. So, it would need some rework (that I'm not to keen on undertaking - this widget is not my best friend :) 

But I would welcome it in the onHold menu, just because it would allow splitting the current 3 buttons row `| Open with... | Convert |Book information|` (where `Book information` is a bit constrained - and truncated to  `|Book informatio|` on my Android) into 2 rows, with a cleaner logical
`| Book status | Book information |`

poire-z commented:
To answer https://www.mobileread.com/forums/showthread.php?t=309179:
> is there any way to change (increase) the cover size in detailed list display modes?

That would be increasing the `64 `there:
https://github.com/koreader/koreader/blob/d4ebcd26c9d146f91be6241cb6bae6501f2cbc92/plugins/coverbrowser.koplugin/listmenu.lua#L736-L739
(and `Display mode > Book info cache management > Delete cache database` or + button > Extract and cache > Refresh already extracted - to get rid of the old sized cached images)

zivilyyyn commented:
It seems that all .mobi files in my library do not show covers in mosaic mode.

Frenzie commented:
It's great that you searched the existing issues first! :-) Unfortunately adding a message to an old, closed issue or PR is at risk of being lost, so in that case it's better to open a new one.

That will be either a crengine or mobi limitation. If the mobi format has a way to indicate cover images like EPUB, I assume it's probably crengine. It it just puts the cover on the first page, it's mobi.

zivilyyyn commented:
Thank you for your reply. Next time I will open a new issue.

For the mobi books, I think I should convert all of them to epub. ;-p

Frenzie commented:
And of course there's also the third option: neither crengine nor mobi are at fault, but your specific books. ;-)
